filter.go

```diff
@@ -0,0 +1,39 @@
+package aiven
+
+import "errors"
+
+var (
+	_ Filter = (*BasicFilter)(nil)
+
+	// ErrUnableToFilter is the error message when the given elements cannot be filtered.
+	ErrUnableToFilter = errors.New("unable to filter given elements")
+)
+
+type (
+	// FilterableElements is a type that wraps the elements that can be filtered.
+	FilterableElements interface{}
+
+	// FiltrationResult is a type that wraps the result of filtering.
+	FiltrationResult interface{}
+
+	// FilterImplementation is a type that implements the Filter interface.
+	FilterImplementation struct{}
+
+	// Filter is a generic filter interface.
+	Filter interface {
+		// Filter filters the given elements.
+		Filter(fe FilterableElements) (FiltrationResult, error)
+	}
+
+	// BasicFilter is a basic filter implementation.
+	BasicFilter FilterImplementation
+)
+
+// Filter filters the given elements.
+func (f *BasicFilter) Filter(fe FilterableElements) (FiltrationResult, error) {
```

Comment:
Since filter has nothing to mutate inside the receiver struct I propose we change it from the pointer to value.
From `func (f *BasicFilter)` to `func (f BasicFilter)`

Reply:
this is going to negatively impact the performance as we will be spawning a new instance of the struct each time the `Filter` method is called

I, however, decided to change the overall behavior of filters and included a preinitialized instances of filter structs, so that we don't need to create them every time we want to filter something out

Reply:
Check these links: 
https://go.dev/doc/faq#methods_on_values_or_pointers

> For types such as basic types, slices, and small structs, a value receiver is very cheap so unless the semantics of the method requires a pointer, a value receiver is efficient and clear.


And:
https://github.com/golang/go/wiki/CodeReviewComments#receiver-type 

> If the receiver is a small array or struct that is naturally a value type (for instance, something like the time.Time type), with no mutable fields and no pointers, or is just a simple basic type such as int or string, a value receiver makes sense. A value receiver can reduce the amount of garbage that can be generated; if a value is passed to a value method, an on-stack copy can be used instead of allocating on the heap. (The compiler tries to be smart about avoiding this allocation, but it can't always succeed.) Don't choose a value receiver type for this reason without profiling first.

Our filter implementation has tiny structures without any internal fields that can be mutated therefore I think we should change this.

Reply:
this is no longer relevant as the implementation has been changed